docs/markdown/Writing Plugins/common-plugin-tasks/plugin-upgrade-guide.md

```diff
@@ -11,36 +11,49 @@ updatedAt: "2022-07-25T20:02:17.695Z"
 
 ### `EnvironmentName` is now required to run processes, get environment variables, etc
 
-In order to support the new [environments](doc:environments) feature, an `EnvironmentName` parameter
-is now required in order to:
+Pants 2.15 introduces the concept of ["Target Environments"](doc:environments), which allow Pants to execute processes in remote or local containerized environments (using Docker), and to specify configuration values for those environments. 
+
+In order to support the new environments feature, an `EnvironmentName` parameter is now required in order to:
 * Run a `Process`
 * Get environment variables
 * Inspect the current `Platform`
 
-Some deprecations are in place to reduce the burden on `@rule` authors:
+This parameter is often provided automatically from a transitive value provided earlier in the call graph. The choice of whether to use a local or alternative environment must be made at a `@goal_rule` level.
+
+In many cases, the local execution environment is sufficient. If so, your rules will not require significant work to migrate, and execution will behave similarly to pre-2.15 versions of Pants.
+
+In cases where the environment needs to be factored into to rule execution, you'll need to do some work. 
+
+2.15 adds a deprecation warning for all goals that have not considered whether they need to use the execution environment.
 
 #### `Goal.environment_behavior`
 
-The `Goal` class has an `environment_behavior` property, which controls whether an `EnvironmentName` is automatically injected when a `@goal_rule` runs. When `environment_behavior=UNMIGRATED` (the default), the `QueryRule` that is installed for a `@goal_rule` will include an `EnvironmentName` and will raise a deprecation error.
+2.15 adds the `environment_behavior` property to the `Goal` class, which controls whether an `EnvironmentName` is automatically injected when a `@goal_rule` runs. 
+
+When `environment_behavior=UNMIGRATED` (the default), the `QueryRule` that is installed for a `@goal_rule` will include an `EnvironmentName` and will raise a deprecation warning.
```

Comment:
these enum values should probably be copy-pastable, meaning that you have the enum class included. Maybe mention the import path too

Reply:
ditto the other references in the file

Reply:
Ack. Happily everyone will have the import class already imported (woo nested namespaces), but I can definitely add the longer test.